grudge/op.py

```diff
@@ -478,11 +481,9 @@ def get_ref_mass_mat(out_grp, in_grp):
 
 def _apply_mass_operator(
         dcoll: DiscretizationCollection, dd_out, dd_in, vec):
-    if isinstance(vec, np.ndarray):
-        return obj_array_vectorize(
-            lambda vi: _apply_mass_operator(dcoll,
-                                            dd_out,
-                                            dd_in, vi), vec
+    if not isinstance(vec, DOFArray):
+        return rec_map_dof_array_container(
```

Comment:
```suggestion
        return map_array_container(
```
There's a lot of traversal functions around these days :\\. The `rec_` versions will basically recurse through the array container themselves until they hit a non-array container. The `dof_array_` versions recurse until they hit a `DOFArray` **or** a non-array container.

For this code, it would be better to
* use the non-recursive `map_dof_array_container`, since it's calling the function again itself, i.e. implementing the recursion right here.
* use the standard `map_array_container`, since it's checking for `DOFArray` itself.
* \+ `map_array_container` is probably the fastest of the bunch, since it has to do the least amount of work.

This should apply to the other changes below too.

Reply:
Good catch, I've replaced all these recursive calls with their non-recursive counterpart: https://github.com/inducer/grudge/commit/18b8c70579801101fb01410b930a4d276a0dafcd